File: 966-binary-subarrays-with-sum/binary-subarrays-with-sum.java
class Solution {
    public int numSubarraysWithSum(int[] nums, int goal) {
        HashMap<Integer, Integer> map = new HashMap<>();
        map.put(0,1);

        int prefix_sum = 0;
        int ans = 0;

        for (int ele : nums) {

            prefix_sum += ele;
            int find = prefix_sum - goal;

            if (map.containsKey(find)) {
                ans += map.get(find);
            }
            
            if (map.containsKey(prefix_sum)) {
                map.put(prefix_sum, map.get(prefix_sum) + 1);
            } else {
                map.put(prefix_sum, 1);
            }
        }

        return ans;
    }
}

File: 966-binary-subarrays-with-sum/binary-subarrays-with-sum.py
class Solution(object):
    def numSubarraysWithSum(self, nums, goal):
        
        map = {}
        map[0] = 1

        ans = 0
        prefix = 0

        for ele in nums:

            prefix += ele

            if (prefix - goal) in map:
                ans += map[prefix - goal]
            
            if prefix in map:
                map[prefix] += 1
            else:
                map[prefix] = 1
        
        return ans
        